graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/json/JsonUtils.java

```diff
@@ -16,11 +16,116 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
 
 public class JsonUtils {
+    private static final Logger LOG = LoggerFactory.getLogger(JsonUtils.class);
+    private static final String KEY_SEPARATOR = "_";
+    private static final String KEY_VALUE_SEPARATOR = ":";
+    private static final String LIST_SEPARATOR = ",";
+    private static final RemoveNullPredicate<Map.Entry<String, Object>> REMOVE_NULL_PREDICATE = new RemoveNullPredicate<>();
+
+    private JsonUtils() {
+    }
+
+    public static JsonNode extractJson(
+            String value, ObjectMapper mapper, boolean flattenObjects, boolean escapeArrays, boolean deleteArrays)
+            throws IOException {
+        if (isNullOrEmpty(value)) {
+            throw new IOException("null result");
+        }
+        final Map<String, Object> json = mapper.readValue(value, TypeReferences.MAP_STRING_OBJECT);
+
+        ObjectNode resultRoot = mapper.createObjectNode();
+        for (Map.Entry<String, Object> mapEntry : json.entrySet()) {
+            for (Entry entry : parseValue(mapEntry.getKey(), mapEntry.getValue(), mapper, flattenObjects, escapeArrays, deleteArrays)) {
+                resultRoot.put(entry.key(), entry.value().toString());
+            }
+        }
+        return resultRoot;
+    }
+
+    private static Collection<Entry> parseValue(
+            String key, Object value, ObjectMapper mapper, boolean flattenObjects, boolean escapeArrays, boolean deleteArrays)
+            throws JsonProcessingException {
+        if (value instanceof Boolean || value instanceof Number || value instanceof String) {
+            return Collections.singleton(Entry.create(key, value));
+        } else if (value instanceof Map) {
+            return parseObject(key, value, mapper, flattenObjects, escapeArrays, deleteArrays);
+        } else if (value instanceof List) {
+            return parseArray(key, value, mapper, flattenObjects, escapeArrays, deleteArrays);
+        } else if (value == null) {
+            // Ignore null values, so we don't try to create fields for that in the message.
+            return Collections.emptySet();
+        } else {
+            LOG.debug("Unknown type \"{}\" in key \"{}\"", value.getClass(), key);
+            return Collections.emptySet();
+        }
+    }
+
+    private static Collection<Entry> parseObject(
+            String key, Object value, ObjectMapper mapper, boolean flattenObjects, boolean escapeArrays, boolean deleteArrays)
+            throws JsonProcessingException {
+        @SuppressWarnings("unchecked")
+        final Map<String, Object> mapWithoutNull = Maps.filterEntries((Map<String, Object>) value, REMOVE_NULL_PREDICATE);
```

Comment:
It would be safer to cast the Object directly after the `instanceof `check and adapt the method signature to: 
`parseObject(
            String key, Map<String, Object> value, ObjectMapper mapper, boolean flattenObjects, boolean escapeArrays, boolean deleteArrays)`